README.md
# Joi-Plus

[![npm package](https://img.shields.io/npm/v/joi-plus.svg)](http://npmjs.org/package/joi-plus) [![npm license](https://img.shields.io/npm/l/joi-plus)](https://img.shields.io/npm/l/joi-plus) [![install size](https://packagephobia.now.sh/badge?p=joi-plus@1.3.0)](https://packagephobia.now.sh/result?p=joi-plus@1.3.0) [![dependencies status](https://david-dm.org/flamehamster/joi-plus/status.svg)](https://david-dm.org/flamehamster/joi-plus)

[joi - v17.4.0](https://www.npmjs.com/package/joi)
Making the most powerful schema description language and data validator for JavaScript slightly more powerful.

## Introduction

* Joi.string().escape()
	* replace `&`, `>`, `<`, `"`, `'`, `\`, `/` and `` ` `` with HTML entities.

* Joi.string().unescape()
	* replace `&amp;` | `&gt;` | `&lt;` | `&quot;` | `&#36;` | `&#47;` | `&#92;` | `&#96;` HTML entities with characters.

* Joi.string().sanitize(function)
	* sanitize string using function that takes a string as a parameter.
	* returns sanitize string

* Joi.string().alpha()
	* Requires the string value to only contain alphabetic characters.

* Joi.string().numeric()
	* Requires the string value to only contain numeric characters.

* Joi.string().decimal(digit, decimal)
	* Requires the string value to be a valid decimal number.

* Joi.string().base32()
	* Requires the value to be a valid base32 string.

* Joi.string().countryCode(type)
	* Requires the value to be a valid ISO `alpha-2` or ISO `alpha-3` country code.

* Joi.string().password(policy)
	* Requires the string value to match policy.
		* policy.min - password minimum length, default 8.
		* policy.max - password maximum length, default 24.
		* policy.lowercase - if true, password requires lowercase.
		* policy.uppercase - if true, password requires uppercase.
		* policy.number - if true, password requires number.
		* policy.special - if true, password requires special character.
		* policy.count
			* If defined, password is required to pass the number of requirements.
			* If undefined, password is required to pass all of requirements.

* Joi.string().match(reference)
	* Requires the string value to match the reference.
	* Removed after validation.

* Joi.string().contain(seed, [index])
	* Requires the string value to contain the seed.
	* If index is defined, position of the seed in the string must match the index.
	* Set index to -1 to match from end of string.

## Quick Start

### Installation
```bash
$ npm i joi-plus
```

### Initialization
```js
const Joi = require('joi-plus');
```

## Usage

```js
const schema = Joi.object({
	username: Joi.string()
		.min(8)
		.max(20)
		.alpha()
		.required(),

	email: Joi.string()
		.email()
		.required(),

	password: Joi.string()
		.password({
			min: 8,
			max: 120,
			lowercase: true,
			uppercase: true,
			number: true,
			special: true,
			count: 3
		})
		.required(),

	repeat_password: Joi.string()
		.match('password')
		.required(),

	base32_encoded: Joi.string()
		.base32()
		.required(),

	country: Joi.string()
		.countryCode('alpha-2')
		.required(),

	contact_number: Joi.string()
		.min(2)
		.max(20)
		.numeric()
		.required(),

	salary: Joi.string()
		.decimal(11,2)
		.required()
});
```

The above schema defines the following constraints:
* `username`
	* a required string
	* at least 8 characters long but no more than 20
	* must contain only alphabetic characters
* `email`
	* a required string
	* a valid email address string
* `password`
	* a required string
	* at least 8 characters long but no more than 120
	* must contain at least 3 of the 4 requirements
		* one lowercase character
		* one uppercase character
		* one numeric character
		* one special character
			* _space_ ! " # $ % & ' ( ) * + , - . : ; < = > ? @ [ \ ] ^ _ \` { | } ~ 
* `repeat_password`
	* a required string
	* must match `password`
	* will be removed after validation
* `base32_encoded`
	* a required string
	* a valid base32 string
* `country`
	* a required string
	* must be a valid ISO 'alpha-2' country code
* `contact_number`
	* a required string
	* at least 2 characters long but no more than 20
	* must contain only numeric characters

* `salary`
	* a required string
	* must be a valid decimal number with up to 11 digits with 2 decimal places

#### Sanitize
Using Joi.string().sanitize() with sanitization libraries such as [sanitize-html](https://www.npmjs.com/package/sanitize-html)

```js
const sanitizeHtml = require('sanitize-html');

const schema = Joi.object({
	escape: Joi.string()
		.escape(),

	unescape: Joi.string()
		.unescape(),

	sanitize: Joi.string()
		.sanitize(sanitizeHtml)
});

const { error, value } = schema.validate({
	escape: '<escape>',
	unescape: '&lt;unescape&gt;',
	sanitize: 'Hello,<script>evil()</script> I am Good.'
});

console.log(value);
/*
{
	escape: '&lt;escape&gt;',
	unescape: '<unescape>',
	sanitize: 'Hello, I am Good.'
}
*/
```

CHANGELOG.md
## Changelog

### 1.4.0

Update [joi](https://www.npmjs.com/package/joi) to version 17.6.0
Added TypeScript support

### 1.3.0

Update [@hapi/joi](https://www.npmjs.com/package/@hapi/joi) version 17.1.1 to [joi](https://www.npmjs.com/package/joi) version 17.4.0

### 1.2.2

Add decimal string validation

### 1.2.1

Update [@hapi/joi](https://www.npmjs.com/package/@hapi/joi) from version 17.1.0 to 17.1.1
Removed array.inList

### 1.2.0

Update [@hapi/joi](https://www.npmjs.com/package/@hapi/joi) from version 16.1.8 to 17.1.0

### 1.1.4

Add requirement count to password policy

* If defined, password is required to pass the number of requirements.
* If undefined, password is required to pass all of requirements.


### 1.1.3

Add contain validation

* `Joi.string().contain('abc')`
* `Joi.string().contain('suffix', -1)`
* `Joi.string().contain('prefix', 0)`

### 1.1.2

Add unescape `&amp;` | `&gt;` | `&lt;` | `&quot;` | `&#36;` | `&#47;` | `&#92;` | `&#96;` HTML entities with characters.

* `Joi.string().unescape()` 

Add sanitize string using function that takes a string as a parameter.

```
Joi.string().sanitize((string) => {
	...
	return string;
})
``` 

### 1.1.1

Fix `type not defined` bug in ISO 3166 `alpha-2` and `alpha-3` country code string validation

### 1.1.0

Update [@hapi/joi](https://www.npmjs.com/package/@hapi/joi) from version 16.1.7 to 16.1.8

Add base32 string validation

* `Joi.string().base32()`

Add ISO 3166 `alpha-2` and `alpha-3` country code string validation

* `Joi.string().countryCode('alpha-2'|'alpha-3')`

Add escape `<`, `>`, `&`, `'`, `"`, `/`, `\` and `` ` `` with HTML entities.

* `Joi.string().escape()`
 
### 1.0.8

Initial commit on github

index.d.ts
import Joi from 'joi';

interface StringSchema extends Joi.StringSchema {
	clean(): this;
	
	/* replace &, >, <, ", ', \, / and ` with HTML entities. */
	escape(): this;

	/* replace &amp;amp; | &amp;gt; | &amp;lt; | &amp;quot; | &amp;#36; | &amp;#47; | &amp;#92; | &amp;#96; HTML entities with characters. */
	unescape(): this;

	/*
		- sanitize string using function that takes a string as a parameter.
		- returns sanitize string
	*/
	sanitize(sanitizer: (value: string) => string): this;

	/* Requires the string value to only contain alphabetic characters. */
	alpha(): this;

	/* Requires the string value to only contain numeric characters. */
	numeric(): this;

	/* Requires the string value to be a valid decimal number. */
	decimal(digit: number, decimal: number): this;

	/* Requires the value to be a valid base32 string. */
	base32(): this;

	/* Requires the value to be a valid ISO alpha-2 or ISO alpha-3 country code. */
	countryCode(type: 'alpha-2' | 'alpha-3'): this;

	password(policy: {
		/* password minimum length, default 8. */
		min?: number;

		/* password maximum length, default 24. */
		max?: number;

		/* if true, password requires lowercase. */
		lowercase?: boolean;

		/* if true, password requires uppercase. */
		uppercase?: boolean;

		/* if true, password requires number. */
		number?: boolean;

		/* if true, password requires special character. */
		special?: boolean;

		/*
			- If defined, password is required to pass the number of requirements.
			- If undefined, password is required to pass all of requirements.
		*/
		count?: number;
	}): this;

	/*
		- Requires the string value to match the reference.
		- Removed after validation.
	*/
	match(reference: string): this;

	/*
		- Requires the string value to contain the seed.
		- If index is defined, position of the seed in the string must match the index.
		- Set index to -1 to match from end of string.
	*/
	contain(seed: string, index?: number): this;
}

interface Root extends Joi.Root {
	string(): StringSchema;
}

declare const root :Root;
export = root;


libraries/string.js
'use strict';
const alpha2 = require('../libraries/alpha-2');
const alpha3 = require('../libraries/alpha-3');

const clean = /^[^><\\\`{}]+$/;
const escape = /^(?=.*[&"'<>`/\\]).*$/;
const unescape = /&amp;|&gt;|&lt;|&quot;|&#36;|&#47;|&#92;|&#96;/;
const alpha = /^[a-zA-Z]+$/;
const numeric = /^[0-9]+$/;
const base32 = /^[A-Z2-7]+=*$/;
const password = /^[ -~]+$/;

const lowercase = /[a-z]/;
const uppercase = /[A-Z]/;
const number = /[\d]/;
const special = /[ -/:-@[-`{-~]/;

module.exports = (joi) => {
	return {
		type: 'string',
		base: joi.string(),
		messages: {
			'string.clean': '{{#label}} contains illegal characters: > < \\ ` } {',
			'string.escape': '{{#label}} contains characters that need to escape: & > < " \' / \\ `',
			'string.unescape': '{{#label}} contains HTML entities that need to unescape: &amp; | &gt; | &lt; | &quot; | &#36; | &#47; | &#92; | &#96;',
			'string.alpha': '{{#label}} must only contain alphabetic characters',
			'string.numeric': '{{#label}} must only contain numeric characters',
			'string.decimal': '{{#label}} must be a decimal number of maximum {{#digit}} digits and {{#decimal}} decimal places',
			'string.base32': '{{#label}} must be a valid base32 string',
			'string.countryCode': '{{#label}} must be a valid ISO {{#type}} country code',
			'string.password': '{{#label}} {{#message}}',
			'string.contain': '{{#label}} must contain "{{#seed}}"{{#message}}'
		},
		coerce(value, helpers) {
			if (helpers.schema.$_getRule('escape')) {
				value = value.replace(/&/g, '&amp;')
							.replace(/>/g, '&gt;')
							.replace(/</g, '&lt;')
							.replace(/"/g, '&quot;')
							.replace(/'/g, '&#36;')
							.replace(/\//g, '&#47;')
							.replace(/\\/g, '&#92;')
							.replace(/`/g, '&#96;');
			}
			if (helpers.schema.$_getRule('unescape')) {
				value = value.replace(/&amp;/g, '&')
							.replace(/&gt;/g, '>')
							.replace(/&lt;/g, '<')
							.replace(/&quot;/g, '"')
							.replace(/&#36;/g, "'")
							.replace(/&#47;/g, '/')
							.replace(/&#92;/g, '\\')
							.replace(/&#96;/g, '`');
			}
			return { value };
		},
		rules: {
			clean: {
				validate: (value, helpers, args, options) => {
					if (clean.test(value)) {
						return value;
					}
					return helpers.error('string.clean');
				}
			},
			escape: {
				convert: true,
				method() {
					return this.$_addRule('escape');
				},
				validate: (value, helpers, args, options) => {
					if (!escape.test(value)) {
						return value;
					}
					return helpers.error('string.escape');
				}
			},
			unescape: {
				convert: true,
				method() {
					return this.$_addRule('unescape');
				},
				validate: (value, helpers, args, options) => {
					if (!unescape.test(value)) {
						return value;
					}
					return helpers.error('string.unescape');
				}
			},
			sanitize: {
				method(sanitizer) {
					return this.$_addRule({ name: 'sanitize', args: { sanitizer } });
				},
				args: [
					{
						name: 'sanitizer',
						assert: (value) => typeof value === 'function',
						message: 'must be a function'
					}
				],
				validate: (value, helpers, args, options) => {
					return args.sanitizer(value);
				}
			},
			alpha: {
				validate: (value, helpers, args, options) => {
					if (alpha.test(value)) {
						return value;
					}
					return helpers.error('string.alpha');
				}
			},
			numeric: {
				validate: (value, helpers, args, options) => {
					if (numeric.test(value)) {
						return value;
					}
					return helpers.error('string.numeric');
				}
			},
			decimal: {
				method(digit, decimal) {
					return this.$_addRule({ name: 'decimal', args: { digit, decimal } });
				},
				args: [
					{
						ref: true,
						name: 'digit',
						assert: (value) => typeof value === 'number',
						message: `must be a number`
					},
					{
						ref: true,
						name: 'decimal',
						assert: (value) => typeof value === 'number',
						message: `must be a number`
					}
				],
				validate: (value, helpers, args, options) => {
					let digit = args.digit;
					let decimal = args.decimal;

					let pattern = `^(0(\\.\\d{1,${decimal}})?|([1-9]\\d{0,${digit - decimal - 1}}(\\.\\d{1,${decimal}})?))$`;
					let regex = new RegExp(pattern);

					if (regex.test(value)) {
						return value;
					}
					return helpers.error('string.decimal', { digit, decimal });
				}
			},
			base32: {
				validate: (value, helpers, args, options) => {
					if (value.length > 0 && value.length % 8 === 0 && base32.test(value)) {
						return value;
					}
					return helpers.error('string.base32');
				}
			},
			countryCode: {
				method(type = 'alpha-2') {
					return this.lowercase().$_addRule({ name: 'countryCode', args: { type } });
				},
				args: [
					{
						ref: true,
						name: 'type',
						assert: (value) => value === 'alpha-2' || value === 'alpha-3',
						message: 'must be "alpha-2" or "alpha-3"'
					}
				],
				validate: (value, helpers, args, options) => {
					let type = args.type;
					if (type === 'alpha-2' && alpha2.includes(value)) {
						return value;
					} else if (type === 'alpha-3' && alpha3.includes(value)) {
						return value;
					}
					return helpers.error('string.countryCode', { type });
				}
			},
			password: {
				method(policy = {}) {
					let min = policy.min || 8;
					let max = policy.max || 24;
					return this.min(min).max(max).$_addRule({ name: 'password', args: { policy } });
				},
				args: [
					{
						name: 'policy',
						assert: (value) => typeof value === 'object' && !Array.isArray(value),
						message: 'must be an object'
					}
				],
				validate: (value, helpers, args, options) => {
					if (!password.test(value)) {
						return helpers.error('string.password', { message: 'contains illegal characters' });
					}

					let policy = args.policy;
					let requirement = 0;
					let count = 0;
					let message = '';
					let error = false;

					if (policy.lowercase) {
						requirement++;
						message += ' one lowercase character,';
						if (lowercase.test(value)) {
							count++;
						}
					}
					if (policy.uppercase) {
						requirement++;
						message += ' one uppercase character,';
						if (uppercase.test(value)) {
							count++;
						}
					}
					if (policy.number) {
						requirement++;
						message += ' one numeric character,';
						if (number.test(value)) {
							count++;
						}
					}
					if (policy.special) {
						requirement++;
						message += ' one special character,';
						if (special.test(value)) {
							count++;
						}
					}
					message = message.slice(0, -1);
					
					if (policy.count && policy.count > 0 && policy.count < requirement) {
						if (policy.count > count) {
							error = true;
							message = ` ${policy.count} of the ${requirement} requirements:` + message;
						}
					} else if (count !== requirement) {
						error = true;
					}

					if (error) {
						message = 'must contains at least' + message;
						return helpers.error('string.password', { message });
					}
					
					return value;
				}
			},
			match: {
				method(reference) {
					return this.valid(joi.ref(reference)).prefs({
						messages: {
							'any.only': `{{#label}} must match "${reference}"`
						}
					}).strip();
				}
			},
			contain: {
				method(seed, index) {
					return this.lowercase().$_addRule({ name: 'contain', args: { seed, index } });
				},
				args: [
					{
						ref: true,
						name: 'seed',
						assert: (value) => typeof value === 'string',
						message: `must be a string`
					},
					{
						ref: true,
						name: 'index',
						assert: (value) => value === undefined || typeof value === 'number' && value >= -1,
						message: `must be equal or more than -1`
					}
				],
				validate: (value, helpers, args, options) => {
					let seed = args.seed;
					let index = args.index;

					if (index) {
						if (index === -1) {
							index = value.length - seed.length;
						}
						if (value.indexOf(seed, index) === index) {
							return value;
						}
						let message = ` at index ${index}`;
						return helpers.error('string.contain', { seed, message });
					} else {
						if (value.includes(seed)) {
							return value;
						}
						return helpers.error('string.contain', { seed });
					}
				}
			}
		}
	}
}